Add human readable conversion of gps time in point clouds

Withalion opened the pull request:
## Description

Point cloud files have GPS time stored for every point, however this value is not easily readable as it is essentially `Double`. This PR adds another attribute field (`GpsTime`) to the point with the original value translated either to the UTC timestamp or day and time in week. The original value of GPS time is moved to `GpsTime (raw)` attribute field.  This feature has been requested in Issue #58695 .

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - you have read the QGIS Developer Guide (https://docs.qgis.org/testing/en/docs/developers_guide/index.html) and your PR complies with its QGIS Coding Standards
-->

Withalion committed:
Add human readable conversion of gps time

nyalldawson commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -291,6 +292,26 @@ struct MapIndexedPointCloudNode
         pointAttr[ QStringLiteral( "Y" ) ] = y;
         pointAttr[ QStringLiteral( "Z" ) ] = z;
+
+        const QDateTime gpsBaseTime = QDateTime::fromSecsSinceEpoch( 315964809, Qt::UTC );
```
Can you bench test this change? My suspicion is that all these datetime calculations will dramatically slow point cloud iteration. (They can be very slow)

Withalion replied:
Do you mean the point cloud iteration during the `identify()` procedure?

I benchmarked this function which is only called while using the identify tool.
Dataset: [Millsite](https://s3.amazonaws.com/data.entwine.io/millsite.copc.laz)
Procedure: I identified 50 points (areas with points) for both versions, started from the overview of the whole point cloud and zoomed in every 10 points. 

|Function| Before change | After change|
| --- | --- | --- |
|Average | 47,4313725490196 ms | 51,843137254902 ms |
| Median | 48 ms | 45 ms |
| Average worst 5% | 104,6 ms | 131 ms |
| Average best 5% | 12,6 ms | 12,8 ms |

wonder-sk commented on src/core/pointcloud/qgscopcpointcloudindex.cpp:
```diff
@@ -408,4 +408,9 @@ QByteArray QgsCopcPointCloudIndex::fetchCopcStatisticsEvlrData()
 
+bool QgsCopcPointCloudIndex::gpsTimeFlag() const
+{
+  return mLazInfo.get()->header().global_encoding % 2;
```
While this is correct, it is preferable to use bitwise AND operator to make it explicit that we want the first bit:
```suggestion
  return mLazInfo.get()->header().global_encoding & 1;
```

wonder-sk commented on src/core/pointcloud/qgscopcpointcloudindex.h:
```diff
@@ -88,6 +88,12 @@ class CORE_EXPORT QgsCopcPointCloudIndex: public QgsPointCloudIndex
 
+    /**
+     * Returns the gps time flag from LAS header
+     * \since QGIS 3.34.12
```
```suggestion
     * \since QGIS 3.42
```

wonder-sk commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -291,6 +292,26 @@ struct MapIndexedPointCloudNode
+
+        const QDateTime gpsBaseTime = QDateTime::fromSecsSinceEpoch( 315964809, Qt::UTC );
+        const int numberOfSecsInWeek = 3600 * 24 * 7;
+        const QgsCopcPointCloudIndex *recastedIndex = dynamic_cast<QgsCopcPointCloudIndex *>( mIndex );
```
maybe `copcIndex` would be a better name

wonder-sk commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -291,6 +292,26 @@ struct MapIndexedPointCloudNode
         pointAttr[ QStringLiteral( "Z" ) ] = z;
+
+        const QDateTime gpsBaseTime = QDateTime::fromSecsSinceEpoch( 315964809, Qt::UTC );
+        const int numberOfSecsInWeek = 3600 * 24 * 7;
```
`gpsBaseTime` and `numberOfSecsInWeek` could be moved inside the "if" below

wonder-sk commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -291,6 +292,26 @@ struct MapIndexedPointCloudNode
+        const int numberOfSecsInWeek = 3600 * 24 * 7;
+        const QgsCopcPointCloudIndex *recastedIndex = dynamic_cast<QgsCopcPointCloudIndex *>( mIndex );
+        if ( recastedIndex )
+        {
```
I would suggest to add a comment what's going on here, and what are the two different meanings of the gps time (+the fact we are also handling the case where the flag is not set, but the number of seconds is high enough)

wonder-sk commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -291,6 +292,26 @@ struct MapIndexedPointCloudNode
+        const QgsCopcPointCloudIndex *recastedIndex = dynamic_cast<QgsCopcPointCloudIndex *>( mIndex );
+        if ( recastedIndex )
+        {
+          const uint16_t gpsTimeFlag = recastedIndex->gpsTimeFlag();
```
```suggestion
          const bool gpsTimeFlag = recastedIndex->gpsTimeFlag();
```

wonder-sk commented on src/core/pointcloud/qgspointclouddataprovider.cpp:
```diff
@@ -291,6 +292,26 @@ struct MapIndexedPointCloudNode
+          const uint16_t gpsTimeFlag = recastedIndex->gpsTimeFlag();
+          if ( gpsTimeFlag || pointAttr[QStringLiteral( "GpsTime" )].toDouble() > numberOfSecsInWeek )
+          {
+            QString utcTime = gpsBaseTime.addSecs( pointAttr[QStringLiteral( "GpsTime" )].toDouble() + qPow( 10, 9 ) ).toString( Qt::ISODate );
```
```suggestion
            QString utcTime = gpsBaseTime.addSecs( pointAttr[QStringLiteral( "GpsTime" )].toDouble() + 1e9 ).toString( Qt::ISODate );
```

wonder-sk commented on src/core/pointcloud/qgspointcloudindex.h:
```diff
@@ -36,6 +36,7 @@
 #include "qgspointcloudattribute.h"
 #include "qgspointcloudexpression.h"
 #include "qgspointcloudrequest.h"
+#include "qgslazinfo.h"
```
this should not be needed

wonder-sk reviewed (commented):
Good stuff... it may be a good idea to move the code to `QgsMapToolIdentify::fromPointCloudIdentificationToIdentifyResults()` so that it is there together with class name handling and adjusted Z handling.

Withalion committed:
Incorporate PR feedback

Withalion commented:
I changed the stuff @wonder-sk mentioned higher

> it may be a good idea to move the code to `QgsMapToolIdentify::fromPointCloudIdentificationToIdentifyResults()` so that it is there together with class name handling and adjusted Z handling

regarding this suggestion, I looked into the code. It would be nice to have it together with the other changes however we would need to access the `QgsPointCloudIndex` through the data provider to get the GPS time flag from the file header and it kinda starts to become a spaghetti

Withalion committed:
Improve readability of if statement

Withalion committed:
Fix copcprovider tests and clang-tidy warning

wonder-sk commented:
Thanks @Withalion!
